kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -414,260 +415,159 @@ public static Tuple2<Integer, Integer> minimumRequiredVersions(Set<TableFeature>
+  /** Utility method to check if the table with given protocol is readable by the Kernel. */
+  public static void validateKernelCanReadTheTable(Protocol protocol, String tablePath) {
+    if (protocol.getMinReaderVersion() > TABLE_FEATURES_MIN_READER_VERSION) {
+      throw DeltaErrors.unsupportedReaderProtocol(tablePath, protocol.getMinReaderVersion());
+    }
 
-  public static final String INVARIANTS_FEATURE_NAME = "invariants";
+    Set<TableFeature> tableSupportedFeatures =
+        Stream.concat(
+                // implicit supported features
+                TABLE_FEATURES.stream()
+                    .filter(
+                        f ->
+                            supportsReaderFeatures(protocol.getMinReaderVersion())
+                                && f.minReaderVersion() <= protocol.getMinReaderVersion()),
+                // explicitly supported features
+                protocol.getReaderFeatures().stream().map(TableFeatures::getTableFeature))
+            .collect(toSet());
 
-  ////////////////////
-  // Helper Methods //
-  ////////////////////
+    Set<TableFeature> unsupportedFeatures =
+        tableSupportedFeatures.stream().filter(f -> !f.hasKernelReadSupport()).collect(toSet());
 
-  public static void validateReadSupportedTable(Protocol protocol, String tablePath) {
-    switch (protocol.getMinReaderVersion()) {
-      case 1:
-        break;
-      case 2:
-        break;
-      case 3:
-        Set<String> readerFeatures = protocol.getReaderFeatures();
-        if (!SUPPORTED_READER_FEATURES.containsAll(readerFeatures)) {
-          Set<String> unsupportedFeatures = new HashSet<>(readerFeatures);
-          unsupportedFeatures.removeAll(SUPPORTED_READER_FEATURES);
-          throw DeltaErrors.unsupportedReaderFeature(tablePath, unsupportedFeatures);
-        }
-        break;
-      default:
-        throw DeltaErrors.unsupportedReaderProtocol(tablePath, protocol.getMinReaderVersion());
+    if (!unsupportedFeatures.isEmpty()) {
+      throw unsupportedReaderFeatures(
+          tablePath, unsupportedFeatures.stream().map(TableFeature::featureName).collect(toSet()));
     }
   }
 
   /**
-   * Utility method to validate whether the given table is supported for writing from Kernel.
-   * Currently, the support is as follows:
-   *
-   * <ul>
-   *   <li>protocol writer version 1.
-   *   <li>protocol writer version 2 only with appendOnly feature enabled.
-   *   <li>protocol writer version 7 with {@code appendOnly}, {@code inCommitTimestamp}, {@code
-   *       columnMapping}, {@code typeWidening}, {@code domainMetadata}, {@code rowTracking} feature
-   *       enabled.
-   * </ul>
-   *
-   * @param protocol Table protocol
-   * @param metadata Table metadata
+   * Utility method to check if the table with given protocol and metadata is writable by the
+   * Kernel.
    */
-  public static void validateWriteSupportedTable(
+  public static void validateKernelCanWriteToTable(
```

Comment:
Can this just call the `validateKernelCanReadToTable` and only check writer features in it?

Reply:
yeah sure.